docs/scos/dev/architecture/architecture-performance.md

```diff
@@ -0,0 +1,211 @@
+
+Some of the Pessimistic locking use cases:
+
+- Concurrent Session problems (Redis, File)
+- Generate Unique Numbers (Database)
+
+Architect can recommend to use Optimistics locking with several different implementations according to the problems.
+
+### Synchronous Communications and 3rd Party Calls
+Another mistake for architecture is to rely on the 3rd party response time in order to achieve a promised performances for the application. Having a direct external call to a 4rd party organisation during a transaction can make the performance unpredictable and will impact the user experience behaviour.
+
+Architects needs to fulfil the requirements by providing a different solution like asynchronous communication or pre/after indirect hooks which are not visible for the end users.
+
+### Native Array vs. DTO
+During a large scale data processing operations, it’s possible to see the performance drawbacks when it comes to object creations, hydrations and mappings in the process. it’s very recommended that architects  use the native language data structure instead of objects if it’s possible, this can reduce more than 30%-50% of CPU loads in long run.
+
+## Performance Optimization in Spryker Architecture
+In this part we would like to analyse the Spryker architecture and find the solutions for the most performance challenges that we had in several projects.
+
+### Database Queries in Plugins
+Spryker uses widely plugins to reduce the module dependencies and increase the flexibility of making features working together, but this can easily lead to some performance issues if we have database queries in each plugin. This is very important to aggregate all queries to decrease the number of database operations.
+
+Example:
+
+We have 10 plugins for cart feature to calculate the items price, discount, tax, etc and each plugin has a query to find product by SKU per order item, it means, the code will execute 10 same queries per each item in cart
+
+If the cart item are equal to 70, we will have 70 x 10 (Plugins) = 700 same queries 
+```
+SELECT * FROM SPY_PRODUCT_ABSTRACT WHERE SKU = ?
+->
+Plugin 1. QUERY
+Plugin 2. QUERY
+Plugin 3. QUERY
+Plugin 4. QUERY
+Plugin n ....
+```
+We can solve this wrong design by
+
+Using IN condition instead of = in query
+```
+SELECT * FROM SPY_PRODUCT_ABSTRACT WHERE SKU IN (?,?,?,....)
+```
+Run only 1 query and provide the result to other plugins
+```
+Plugin 1. QUERY
+Plugin 2. RESULT
+Plugin 3. RESULT
+Plugin 4. RESULT
+Plugin n ...
+```
+
+### ORM vs PDO
+Spryker uses Propel ORM to reduce the database complexity and low level interactions for the developers, as this is very great abstraction and cleaner approach but developers need to be careful to not miss use the pattern so some specific places such as heavy data transfer operations like Data Import or P&S for big data.
```

Comment:
I don't quite understand what *as this is very great abstraction and cleaner approach* means here. Could you please elaborate a bit?

Reply:
Spryker uses Propel ORM as a database abstraction layer which allows to stay DBMS agnostic and use its powerful tools for persisting and accessing data in a clean and readable way. But this comes with some performance costs which can be especially noticeable when working with heavy data transfer operations like Data Import and P&S for big data volumes.